### PR TITLE
Adding Acceptance Tests for Translation Layer

### PR DESCRIPTION
**Pull request by khilangoel**
Refer : https://github.com/Microsoft/vstest/issues/1239

### CODE REVIEW

File: test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs [hunk 1]
```diff
@@ -95,5 +97,34 @@ protected string GetTestHostProcessName(string targetPlatform)
...
+
+        protected IVsTestConsoleWrapper GetVsTestConsoleWrapper()
+        {
+            var vsConsoleWrapper = new VsTestConsoleWrapper(this.GetConsoleRunnerPath());
+            vsConsoleWrapper.StartSession();
+
+            return vsConsoleWrapper;
+        }
+
+        public string GetDefaultRunSettings()
```
Comment (codito) [inline test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs]: Require docs for public APIs.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class RunSelectedTests : AcceptanceTestBase
+    {
+        [TestMethod]
+        public void RunSelectedTestsWithoutTestPlatformOptions()
+        {
+            var sources = new List<string>
+                              {
```
Comment (codito) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs]: DRY

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ public interface IVsTestConsoleWrapper
         /// <summary>
         /// Initialize the TestPlatform with Paths to extensions like adapters, loggers and any other extensions
         /// </summary>
-        /// <param name="pathToAdditionalExtensions">Folder Paths to where extension DLLs are present</param>
+        /// <param name="pathToAdditionalExtensions">Full Paths to where extension DLLs are present</param>
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs]: Yes. Will change it.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            Assert.AreEqual(6, this.discoveryEventHandler2.DiscoveredTestCases.Count);
+            Assert.IsTrue(this.discoveryEventHandler2.Metrics.ContainsKey(TelemetryDataConstants.TargetDevice));
+            Assert.IsTrue(this.discoveryEventHandler2.Metrics.ContainsKey(TelemetryDataConstants.NumberOfAdapterUsedToDiscoverTests));
+            Assert.IsTrue(this.discoveryEventHandler2.Metrics.ContainsKey(TelemetryDataConstants.TimeTakenInSecByAllAdapters));
+            Assert.IsTrue(this.discoveryEventHandler2.Metrics.ContainsKey(TelemetryDataConstants.TimeTakenInSecForDiscovery));
+            Assert.IsTrue(this.discoveryEventHandler2.Metrics.ContainsKey(TelemetryDataConstants.DiscoveryState));
+        }
+
+        [TestMethod]
+        public void DiscoverTestsUsingEventHandler2AndBatchSize()
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs]: In both, RunSettings are different.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        [TestMethod]
+        public void DiscoverTestsUsingEventHandler1AndBatchSize()
+        {
+            this.vstestConsoleWrapper.DiscoverTests(
+                this.testAssemblies,
+                this.runSettingsXml,
+                this.discoveryEventHandler);
+
+            // Assert.
+            Assert.AreEqual(6, this.discoveryEventHandler.DiscoveredTestCases.Count);
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs]: Not supported yet

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        private IVsTestConsoleWrapper vstestConsoleWrapper;
+        private List<string> testAssemblies;
+        private RunEventHandler runEventHandler;
+        private DiscoveryEventHandler discoveryEventHandler;
+
+        public RunSelectedTests()
+        {
+            this.testAssemblies = new List<string>
+                              {
+                                  this.GetAssetFullPath("SimpleTestProject.dll"),
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs]: Done

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                                  this.GetAssetFullPath("SimpleTestProject2.dll")
+                              };
+
+            this.vstestConsoleWrapper = this.GetVsTestConsoleWrapper();
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunSelectedTests.cs]: Will do in next commit

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+                this.GetDefaultRunSettings(),
+                new TestPlatformOptions() { TestCaseFilter = "FullyQualifiedName=SampleUnitTestProject.UnitTest1.PassingTest" },
+                this.runEventHandler);
+
+            // Assert
+            Assert.AreEqual(1, this.runEventHandler.TestResults.Count);
+            Assert.AreEqual(TestOutcome.Passed, this.runEventHandler.TestResults.FirstOrDefault().Outcome);
+
+            // Release builds optimize code, hence line numbers are different.
+            if (IntegrationTestEnvironment.BuildConfiguration.StartsWith("release", StringComparison.OrdinalIgnoreCase))
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs]: Done

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        [TestMethod]
+        public void DiscoverTestsUsingDiscoveryEventHandler2AndTelemetryOptedOut()
+        {
+            this.vstestConsoleWrapper.DiscoverTests(
+                this.testAssemblies,
+                this.GetDefaultRunSettings(),
+                new TestPlatformOptions() { CollectMetrics = false },
+                this.discoveryEventHandler2);
+
+            // Assert.
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs]: Done.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+
+            var sources = new List<string>
+                              {
+                                  this.GetAssetFullPath("NUTestProject.dll")
+                              };
+
+            this.vstestConsoleWrapper.RunTests(
+                sources,
+                this.GetDefaultRunSettings(),
+                new TestPlatformOptions() { TestCaseFilter = "FullyQualifiedName=NUnitTestProject.NUnitTest1.PassTestMethod1" },
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs]: Run Failed test too.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+                              };
+
+            this.vstestConsoleWrapper.RunTests(
+                sources,
+                this.GetDefaultRunSettings(),
+                new TestPlatformOptions() { TestCaseFilter = "FullyQualifiedName = SampleUnitTestProject3.UnitTest1.WorkingDirectoryTest" },
+                this.runEventHandler);
+
+            // Assert
+            Assert.AreEqual(1, this.runEventHandler.TestResults.Count);
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs]: Check for testhost.exe.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+        [NETFullTargetFramework]
+        [NETCORETargetFramework]
+        public void RunTestsWithNetCoreProject(RunnerInfo runnerInfo)
+        {
+            AcceptanceTestBase.SetTestEnvironment(this.testEnvironment, runnerInfo);
+            this.ExecuteNotSupportedRunnerFrameworkTests(runnerInfo.RunnerFramework, Netcoreapp, Message);
+
+            var sources = new List<string>
+                              {
+                                  this.testEnvironment.GetTestAsset("SimpleTestProject.dll", "netcoreapp1.0"),
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs]: Explicit targetframework not required.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+        public void RunTestsShouldThrowOnStackOverflowException(RunnerInfo runnerInfo)
+        {
+            AcceptanceTestBase.SetTestEnvironment(this.testEnvironment, runnerInfo);
+            this.ExecuteNotSupportedRunnerFrameworkTests(runnerInfo.RunnerFramework, Netcoreapp, Message);
+
+            if (IntegrationTestEnvironment.BuildConfiguration.Equals("release", StringComparison.OrdinalIgnoreCase))
+            {
+                // On release, x64 builds, recursive calls may be replaced with loops (tail call optimization)
+                Assert.Inconclusive("On StackOverflowException testhost not exited in release configuration.");
+                return;
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/RunTests.cs]: return not required.

**Review by smadala — APPROVED** [review]

**Review by mayankbansal018 — APPROVED** [review]

**Pull request merged**